compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java

```diff
@@ -1273,6 +1274,10 @@ public void visit(BLangSimpleVarRef varRefExpr) {
 
         if (!recordFieldDefaultValue && !objectFieldDefaultValueRequiringIsolation && enclInvokable != null &&
                 isReferenceToVarDefinedInSameInvokable(symbol.owner, enclInvokable.symbol)) {
+            if (this.inIsolatedStartAction
+                    && !isSubtypeOfReadOnlyOrIsolatedObjectOrInferableObject(symbol.owner, symbol.getType())) {
+                inferredIsolated = false;
+            }
```

Comment:
Can't this happen outside the outer check?

Reply:
At the outside of the outer check we analyse symbols with `recordFieldDefaultValue` and `objectFieldDefaultValueRequiringIsolation` being true. For those I don't think we need to check its a subtype of readonly or isolated object.

Reply:
But for those `this.inIsolatedStartAction` won't be true, right?

Reply:
Yes. But outer was there to handle both `true` and `false` scenarios. We have add this new check to validate start actions specifically. Which wasn't being analysed previously.

Reply:
Then we need `isReferenceToVarDefinedInSameInvokable` check as well.

Reply:
When I am running the test locally previously seems I haven't added the `return` statement. We can move this to the outside of the outer if. Updated in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/42366/commits/a22e04b5a5c73c4d6a0327088fd0126d7125d7a5).